File: lib/rest/yampi/mappers/yampi_order_mapper.dart
import 'package:sertton/core/checkout/dtos/order_dto.dart';
import 'package:sertton/rest/types/json.dart';
import 'package:sertton/rest/yampi/mappers/yampi_address_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_order_item_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_payment_mapper.dart';

class YampiOrderMapper {
  static OrderDto toDto(Json json) {
    if (json.containsKey('data')) {
      json = json['data'];
    }

    final statusData = json['status']?['data'];
    final statusAlias = statusData?['alias'] ?? json['status_alias'] ?? '';

    return OrderDto(
      status: _mapStatus(statusAlias),
      number: json['number']?.toString() ?? '',
      shippingName: json['shipment_service'] ?? json['shipping_service'] ?? '',
      shippingPrice:
          double.tryParse(
            (json['value_shipment'] ?? json['price_shipment']).toString(),
          ) ??
          0.0,
      shippingAddress: YampiAddressMapper.toDto(json['shipping_address'] ?? {}),
      items: YampiOrderItemMapper.toDtoList(json['items'] ?? {}),
      payment: YampiPaymentMapper.toDtoFromTransactions(
        json['transactions'] ?? {},
      ),
      createdAt:
          DateTime.tryParse(json['created_at']?['date'] ?? '') ??
          DateTime.now(),
    );
  }

  static List<OrderDto> toDtoList(Json json) {
    if (json.containsKey('data')) {
      final data = (json['data'] as List).cast<Json>();
      return data.map(toDto).toList();
    }
    return [];
  }

  static OrderStatus _mapStatus(String alias) {
    switch (alias) {
      case 'paid':
        return OrderStatus.paid;
      case 'created':
        return OrderStatus.created;
      case 'cancelled':
        return OrderStatus.cancelled;
      case 'refused':
        return OrderStatus.refused;
      case 'authorized':
        return OrderStatus.authorized;
      case 'delivered':
        return OrderStatus.delivered;
      case 'waiting_payment':
        return OrderStatus.waitingPayment;
      default:
        return OrderStatus.created;
    }
  }
}

File: sertton-react-native-app/src/rest/yampi/mappers/yampi-order-mapper.ts
import { type Order, OrderStatus, PaymentMethod } from "@/core/checkout/entities"
import type { YampiOrder } from "@/rest/yampi/types"
import { YampiAddressMapper } from "./yampi-address-mapper"
import { YampiOrderItemMapper } from "./yampi-order-item-mapper"
import { YampiPaymentMapper } from "./yampi-payment-mapper"

export const YampiOrderMapper = () => {
  const addressMapper = YampiAddressMapper()
  const orderItemMapper = YampiOrderItemMapper()
  const paymentMapper = YampiPaymentMapper()

  const unwrap = <Value>(value: Value | { readonly data: Value } | undefined): Value | undefined =>
    value && typeof value === "object" && "data" in value ? value.data : value

  const first = <Value>(
    value: Value | readonly Value[] | { readonly data: Value | readonly Value[] } | undefined,
  ): Value | undefined => {
    const unwrapped = unwrap<Value | readonly Value[]>(value)
    if (Array.isArray(unwrapped)) return unwrapped[0] as Value | undefined
    return unwrapped as Value | undefined
  }

  return {
    toDomain(input: YampiOrder): Order {
      const statusValue = unwrap(input.status)
      const shippingAddress = first(input.shipping_address) ?? first(input.address)
      const items = unwrap(input.items) ?? []
      const transaction = first(input.transactions)
      const status =
        {
          paid: OrderStatus.paid,
          cancelled: OrderStatus.cancelled,
          refused: OrderStatus.refused,
          authorized: OrderStatus.authorized,
          delivered: OrderStatus.delivered,
          waiting_payment: OrderStatus.waitingPayment,
        }[typeof statusValue === "string" ? statusValue : (statusValue?.alias ?? "")] ??
        OrderStatus.created
      const createdAt = new Date(
        typeof input.created_at === "string" ? input.created_at : (input.created_at?.date ?? ""),
      )

      return {
        status,
        number: String(input.number),
        shippingName: input.shipment_service ?? "",
        shippingPrice: input.value_shipment ?? 0,
        shippingAddress: shippingAddress
          ? addressMapper.toDomain(shippingAddress)
          : {
              id: "",
              receiver: "",
              zipcode: "",
              street: "",
              number: "",
              neighborhood: "",
              complement: "",
              city: "",
              uf: "",
            },
        items: items.map(orderItemMapper.toDomain),
        payment: transaction
          ? paymentMapper.toDomain(transaction)
          : { id: "", name: "N/A", icon: "", method: PaymentMethod.boleto },
        createdAt: Number.isNaN(createdAt.valueOf()) ? new Date(0) : createdAt,
      }
    },
  }
}
